rollup/fees/rollup_fee.go

```diff
@@ -48,59 +44,82 @@ type StateDB interface {
 	GetBalance(addr common.Address) *big.Int
 }
 
-// CalculateL1MsgFee computes the L1 portion of the fee given
-// a Message and a StateDB
-// Reference: https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/rollup/fees/rollup_fee.go
-func CalculateL1MsgFee(msg Message, state StateDB) (*big.Int, error) {
-	tx := asTransaction(msg)
+func EstimateL1DataFeeForMessage(msg Message, baseFee, chainID *big.Int, signer types.Signer, state StateDB) (*big.Int, error) {
+	unsigned := asUnsignedTx(msg, baseFee, chainID)
+	tx, err := unsigned.WithSignature(signer, bytes.Repeat([]byte{0xff}, crypto.SignatureLength))
+	if err != nil {
+		return nil, err
+	}
+
 	raw, err := rlpEncode(tx)
 	if err != nil {
 		return nil, err
 	}
 
 	l1BaseFee, overhead, scalar := readGPOStorageSlots(rcfg.L1GasPriceOracleAddress, state)
-	l1Fee := CalculateL1Fee(raw, overhead, l1BaseFee, scalar)
-	return l1Fee, nil
+	l1DataFee := CalculateL1Fee(raw, overhead, l1BaseFee, scalar)
+	return l1DataFee, nil
 }
 
-// asTransaction turns a Message into a types.Transaction
-func asTransaction(msg Message) *types.Transaction {
-	if msg.To() == nil {
-		return types.NewContractCreation(
-			msg.Nonce(),
-			msg.Value(),
-			msg.Gas(),
-			msg.GasPrice(),
-			msg.Data(),
-		)
+// asUnsignedTx turns a Message into a types.Transaction
+func asUnsignedTx(msg Message, baseFee, chainID *big.Int) *types.Transaction {
+	if msg.AccessList() == nil {
```

Comment:
Is it possible that `msg.AccessList() == nil` but `baseFee != nil`?

Reply:
Currently not since EIP-1559 tx also has an access list.

Reply:
Yeah, but if the access list is empty, it might be `nil`.